src/pip/_internal/download.py

```diff
@@ -139,31 +144,78 @@ def __init__(self, prompting=True):
         self.prompting = prompting
         self.passwords = {}
 
-    def __call__(self, req):
-        parsed = urllib_parse.urlparse(req.url)
+    def _get_url_and_credentials(self, original_url,
+                                 allow_netrc=True, allow_keyring=True):
+        """Returns the credentials to use for the provided URL.
+
+        If allowed, netrc and keyring may be used to obtain the
+        correct credentials.
+
+        Returns (url_without_credentials, username, password). Note
+        that even if the original URL contains credentials, the result
+        of this function may include different values.
+        """
+        parsed = urllib_parse.urlparse(original_url)
 
         # Get the netloc without any embedded credentials
         netloc = parsed.netloc.rsplit("@", 1)[-1]
 
-        # Set the url of the request to the url without any credentials
-        req.url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
+        # Return the url without any credentials
+        url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
 
         # Use any stored credentials that we have for this netloc
         username, password = self.passwords.get(netloc, (None, None))
 
         # Extract credentials embedded in the url if we have none stored
         if username is None:
             username, password = self.parse_credentials(parsed.netloc)
+            if username or password:
+                logger.debug("Found credentials in url for %s", netloc)
```

Comment:
Sorry, I didn't meant to suggest a debug log at every branch point -- just ones specific to the keyring if it would make sense and not too verbose. I don't have a good sense though for whether a given message would result in the debug logs being too noisy, but maybe this would be okay. Would these log messages let you reconstruct: 1) which data source the password is coming from for a given netloc, and 2) whether access to the keyring is being attempted? For (2), I'm not sure if the current log messages distinguish between the keyring not being accessed and the keyring being accessed but a password not being found.

Reply:
The only case where more than one of these messages will be displayed is when the URL includes a username but no password, and keyring is able to retrieve the password for that user. So I think it makes sense to identify how the credentials were found in any particular call, and this achieves it.

There's no specific message for keyring being accessed. Just a debug message when it succeeds and a warning when it unexpectedly fails. I'll change them up a bit to be more before/after type messages.